File: JAVA/19thOct-4.java
package Java;

import java.util.Scanner;

/**
 * Program			=>		Implementation of Element Search in Matrix (with rows and Columns Sorted) Issue #530
 * Issue Details	=> 		https://github.com/twowaits/make-pull-request/issues/530
 * Contributed by	=>		@sumitaccess007 : https://github.com/sumitaccess007
 */

public class MatrixSearch
{

	public static void main(String[] args)
	{
		
		int m, n;	// Rows and Columns of Matrix
		
		Scanner scanner = new Scanner(System.in);
		System.out.println("Enter the number of row in matrix :- ");	// Rows Input from User
		m = scanner.nextInt();
		
		System.out.println("Enter the number of columns in matrix :- ");	// Columns Input from User
		n = scanner.nextInt();
		
		int matrix[][] = new int[m][n];
		System.out.println("Enter all the elements of the matrix (Rows and Columnd Sorted in Ascending Order) :- ");	// Requested all the elements of matrix (with rows and Columns Sorted)
		for (int i=0; i<m; i++)
		{
			for (int j=0; j<n; j++)
			{
				matrix[i][j] = scanner.nextInt();
			}
		}
		
		int x;
		System.out.println("Enter the element to search :- ");	// Element to Search
		x = scanner.nextInt();
		
		System.out.println("Final Matrix is :- ");
		for (int i=0; i<m; i++)
		{
			for (int j=0; j<n; j++)
			{
				System.out.print(matrix[i][j]+" ");		// Printed the matrix
			}
			System.out.println("");
		}
		
		// Search for the elemtnt in matrix and Print final result
		System.out.println("Searching Element "+ x +" in Matrix (true - For present and false - Not Present) ...");
		System.out.println(searchInMatrix(matrix, m, n, x));

	}

	// Divide and Conquer Approach
	private static boolean searchInMatrix(int[][] matrix, int m, int n, int x) {
		
		int p=0, q=n-1;	// Start from top right corner
		
		while (p < m && q >= 0)
		{
			if (matrix[p][q] == x)
			{
				return true;	// Element Found
			}
			if (matrix[p][q] < x)
			{
				p++;	// Eliminate Rows if Matrix number if smaller than element to search
			}
			else
			{
				q--;	// Eliminate Columns if Matrix number if larger than element to search
			}
		}
		return false;	// Element Not Found in Matrix

	}
	
}

File: C++/19thOct-4.cpp
/** Contributed by: ADIT SHAH
    Algorithm: Searching of elements in a row-wise and column-wise sorted matrix
**/

#include<iostream>
using namespace std;
int main()
{
    int row,col,x;
    cout<<"Enter the dimensions of the matrix: ";
    cin>>row>>col;
    cout<<"\nEnter the matrix:"<<endl;
    int arr[row][col];
    for(int i=0;i<row;i++)
        for(int j=0;j<col;j++)
            cin>>arr[i][j];
    char ch='y';
    while(ch=='y' || ch=='Y')
    {
        cout<<"\nEnter the element you want to search for: ";
        cin>>x;
        bool isFound=false;
        if(row>0 && col>0)
        {
            int s=arr[0][0];
            int l=arr[row-1][col-1];
            if(x>=s && x<=l)
            {
                int i=row-1,j=0;
                while(i>=0 && j<col)
                {
                    if(x==arr[i][j])
                    {
                        isFound=true;
                        break;
                    }
                    else if(x<arr[i][j])
                        i--;
                    else
                        j++;
                }
            }
        }
        if(isFound)
            cout<<"TRUE\n";
        else
            cout<<"FALSE\n";
        cout<<"Want to search more elements ?:(y or n) ";
        cin>>ch;
    }
    return 0;
}
